### DISCUSSION TITLE
Thumbnail Generation taking forever

### DISCUSSION THREAD

**Discussion by andyxpert** [immich-app/immich#6188] [Q&A]
Hello, I'll try to be brief...
Been using immich for almost a year already and love it.
Recently I'm having an issue when I run the Library job on the external lib (~65k files, mostly photos), after I add some new files to the folder, it takes around half an hour (slow mini-pc) to parse all the files, then it extracts metadata (mostly in parallel) and then it starts generating thumbnails. Nothing off so far.
The problem is that after a while (1-2 hours or so) only the thumbnail job remains active and it seems to be veeeeeeeeery slow.
Slow meaning that it processes around 4-5 items per minute... and having 54k+ files pending it takes close to a week to finish !!

is this normal ?
Why does the thumbnail generation speed drop so much and why does it take that long ?
The microservices node is hogging the RAM, nothing interesting in the logs...

attached a screenshot of the pending taska for the 3 jobs i mentioned - it's over a 12 hour timespan. You can clearly see the spikes when it started the library job (around 10h ago).
![IMG_1305](https://github.com/immich-app/immich/assets/69542335/b525ba36-8316-4de1-8f31-722061391616)

What do you guys think ?

**Comment by zackpollard** [comment]
What do you have your concurrency set to on that job? Setting that value too high can actually be counterproductive. Also what CPU do you have in the mini pc? It's possible it's just not a very fast cpu. Also there are other things that increase the number of thumbnails to be generated as the job continues, like facial recognition.

**Reply by andyxpert** [reply]
An older i5, concurrency set to 2-4 as I don't have many cores :)

I think it's related to the thumbnail structure that was chamged a few months ago and I think my migration didn't complete properly as it all works ok up to a point and then it starts lagging like hell.
Also while the loooooong thumbnail job was running, i saw in the microservices log that it kept on generating, but no new file created which begs the question - what was it doing if NOT generating actual files ?!?

i think that if the migration job is not run properly, or if there are inconsistencies in the db with the thumbnail hashes it enters a loooooong loop where it takes all items one by one and tries to identify existing thumbs, else generates them, but if the file already exists it skips or waits a while... causing such a low speed of job processing.

I'm trying now to re-run the migration job and check for any issues in the logs, then i'll re-run the library job and get back with updates. It may be isolated to an improper existing thumbnail structure that's broken...

**Reply by andyxpert** [reply]
Update: ran the Migration job successfully then ran the Library job - again finished really fast - but I keep seeing in the log some errors about duplicates (exactly this scenario: https://github.com/immich-app/immich/issues/5067 ) 
Will try to re-run the thumbnail job and see how it behaves, now that migration and library jobs ran successfully I expect no issues in the thumbnail folder structure.

**Reply by andyxpert** [reply]
So far it seems that re-running the Migration job fixed the problem with the neverending thumbnails job... Guess there was an inconsistency in the thumbnail folder structure... 
I still see errors in the log about duplicates ("UQ_assets_owner_library_checksum" violation), I am inclined to re-generate all the thumbnails this night... on a stronger machine and move them (with the DB of course) to the low-power server just to ditch the errors... let's see.
I'll try to do the addition flow now (adding a few images/videos to the external lib, then run the library job and see if the new files appear in the timeline AND hope no job will start lagging like crazy).

If all goes well I'll write the conclusion and the discussion can be safely closed

**Reply by andyxpert** [reply]
Back to square 1... 
I added some new media files, ran the Library job - finished, the files appear in the timeline, all good
HOWEVER, back to the original problem (graphs span for last 2 hours):
![image](https://github.com/immich-app/immich/assets/69542335/c9fa315f-763f-4d32-b35d-27f66341b16c)

Logs of microservices node keep on saying it's generating thumbnails (though the filesystem has the same number of files, so no new thumbs generated... or maybe they're overwritten ...) and the constraint violation error in postgres:
ERROR:  duplicate key value violates unique constraint "UQ_assets_owner_library_checksum"
2024-01-05 12:47:24.172 UTC [599] DETAIL:  Key ("ownerId", "libraryId", checksum)=(69301539-900b-4882-bac4-df8f57ab2f47, a9b4cad9-1fd3-4f9c-8e83-cf7d14354ce6, \x8436d0d009a190993150ba448eeebcef196831be) already exists.
2024-01-05 12:47:24.172 UTC [599] STATEMENT:  INSERT INTO "assets"(...

If you suggest anything, I'm all ears.
If not... I'll re-create the entire DB and thumbnails structure on a beefier hardware and then move them to the server.... hopefully starting fresh clears the issues...

PS: why does the Storage Template Migration job keep on starting (every file that's parsed by the metadata extraction) ?? Is there any template migration that's required when simply running the Library job ?

**Comment by andyxpert** [comment]
Conclusion after some testing:
I was calling the library job through the API with force set to true which i guess forces all thumbnails to be regenerated (explains my issue with the looooong job).
Setting it to false, or calling the job from the GUI (Refresh Gallery) didn't refresh/generate the thumbnails also as i thought...

anyway, now i run library refresh and then missing thumbnails job and that seems to do the trick.
I'll probably automate this soon...
A CRON for thumbnail refresh (like the one for library scan) would be nice ;)

**Comment by Nancy-Robertson** [comment]
@andyxpert I have a problem with thumbnail generation as well, but I am totally new to immich so I might have something wrong with my setup.
I was wondering how you get the nice graphs that show the job load?

My external library of ~120,000 photos is mounted over smb, and thubmnail generation never finishes...

I tried using tree to count the images in the thumbs folder, and I have seen it go down, but I do not see anything interesting in the logs.

**Reply by mertalev** [reply]
We had a bug a few releases back where external libraries could import generated images as new images and make an infinite feedback loop. You might be affected if you had an import path that includes the `thumbs` folder (like `.` or `/usr/src/app/upload`), used Immich on 1.97 or 1.98 and had file watching on.

If that doesn't apply, it's probably just that generating so many images takes a good deal of time. Images are also added to the queue as other jobs get processed, so it's normal for the waiting number to increase even as jobs finish because there's a steady stream of new jobs being added.

If you have an SSD and a beefy CPU and notice they both aren't getting fully utilized, you can set the `UV_THREADPOOL_SIZE` env var to a higher number and increase the job concurrency.

Conversely, if you're seeing that everything is fully pinned or you're using a lot of swap memory, you may want to pause the machine learning jobs so the other jobs can get processed more efficiently.

**Reply by 825i** [reply]
Pretty sure I am experiencing this issue. #11040 

Or is there good reason my thumbs folder has 150k images in it for 90k photos in my library?

**Reply by mertalev** [reply]
That's normal. A thumbnail and preview image file is generated for each asset, and person thumbnails also exist in that folder.

**Reply by andyxpert** [reply]
> @andyxpert I have a problem with thumbnail generation as well, but I am totally new to immich so I might have something wrong with my setup. I was wondering how you get the nice graphs that show the job load?

I use Home Assistant to monitor Immich Job status through its API

> My external library of ~120,000 photos is mounted over smb, and thubmnail generation never finishes...

Try clearing the job and restarting the service then restarting the job (for the missing files)

**Comment by andyxpert** [comment]
Update: upgraded my mini-PC, now running a 9th gen i5 6 cores with 32 GB RAM. Allocated 2 cores and 8 GB for Immich Services
Indeed a significant speedup in thumbnail generation, but there is still a problem... 
Started 24h ago the initial ext Library scan, in a few hours most of the jobs finished, thumb generation and metadata remained working. When metadata finished I saw a sudden double in the thumb waiting jobs (from a normal 60k to around 120k and that number constantly grew.
Forced stopped the job, cleared the remaining waiting files and restarted the stack. Restarted the thumb generation (for the missing ones) and got back to normal (60-ish k with the number now decreasing at a normal speed (7-8/s)

I think there is still a bug in the thumb job which gets fixed/avoided by clearing the job cache and restarting the entire stack (or just the service)

Btw, the log is clean no duplicate files or error generating thumbs and the thumbs seem to be constantly generated, so the error might be only on how the remaining job count is calculated (increasing instead of decreasing)

![Screenshot_20240719_231220_edit_116845629577482](https://github.com/user-attachments/assets/86c0bbb1-619b-4f5f-960f-2f2fe26e0b8a)


@alextran1502

**Comment by bhugh** [comment]
FYI I'm having the same problem as @andyxpert just above.  The jobs page reports 3XX,XXX thumbnails "waiting" and even though it is generating thumbnails at a decent clip, the cue just gets ever-larger.  It grew from about 200,000 to around 350,000 in maybe an hour.

I have about 250K images but all the thumbnails, except for a batch of around 1000 photos I just added, had previously been generated & working fine.

Looking at the immich logs, it looks like it is generating new thumbnails for all sorts of images and then just deleting the previously generated thumbnails for the image.

Also - and what brought this to my attention - the thumbnails for the newly added images were not being generated.  They just showed up as and X.

_docker compose down redis_  followed by _docker compose up redis -d_ got things back on track.  It generated missing thumbnails for the new 1000 or so photos in just a few minutes and seems just fine now.

**Comment by kostnerek** [comment]
I'm also having this issue. External library of 32k photos is generating its thumbnails from yesterday at full blast of ryzen 5 5600g + 8gb of RAM. I cannot comprehend why every time i look into it's stats, and stop generating and rerun job for missing thumbs it says different number (I'm not talking about differences like few hundreds, but difference  between 10k-90k). Speeds I'm getting is 3/4 photos per minute at best. My concurent jobs were changed 3-6 which did nothing. Everything is hosted locally on Unraid through docker compose

**Reply by mertalev** [reply]
Did you enable full-size thumbnail generation? That can slow down processing quite a bit. Otherwise, the speed depends on a lot of things. It could be falling back to swap if there isn't enough ram, it could be very slow disk speed, or it could be that your processor is just that slow (granted a 5600g should be much faster than 3-4 assets per minute). If you have a ton of videos, generating thumbnails for them is also more intensive than for photos.
